src/keys/slip10.rs

```diff
@@ -99,384 +188,203 @@ impl Seed {
         Self(bs.to_vec())
     }
 
-    pub fn to_master_key(&self, curve: Curve) -> ExtendedSecretKey {
-        ExtendedSecretKey::from_seed(curve, self)
+    pub fn to_master_key<K: hazmat::IsSecretKey>(&self) -> Extended<K> {
+        Extended::from_seed(self)
     }
 
-    pub fn derive(&self, curve: Curve, chain: &Chain) -> crate::Result<ExtendedSecretKey> {
-        self.to_master_key(curve).derive(chain)
+    pub fn derive<K: hazmat::IsSecretKey>(&self, chain: &Chain) -> crate::Result<Extended<K>> {
+        self.to_master_key().derive(chain)
     }
 }
 
 pub type ChainCode = [u8; 32];
 
-#[derive(Clone, Zeroize, ZeroizeOnDrop)]
-pub struct ExtendedSecretKey(KeyImpl);
-
-impl ExtendedSecretKey {
-    pub fn from_seed(curve: Curve, seed: &Seed) -> Self {
-        Self(KeyImpl::from_seed(curve, &seed.0))
-    }
+pub struct Extended<K> {
+    key: core::marker::PhantomData<K>,
+    ext: [u8; 65],
+}
 
-    pub fn try_from_extended_bytes(curve: Curve, ext_bytes: &[u8; 64]) -> crate::Result<Self> {
-        let mut key = KeyImpl { curve, ext: [0_u8; 65] };
-        key.ext[1..].copy_from_slice(ext_bytes);
-        if key.is_secret_key_valid() {
-            Ok(Self(key))
-        } else {
-            Err(crate::Error::InvalidArgumentError {
-                alg: "SLIP10",
-                expected: "extended secret key",
-            })
+impl<K> Clone for Extended<K> {
+    fn clone(&self) -> Self {
+        Self {
+            key: core::marker::PhantomData,
+            ext: self.ext,
         }
     }
+}
 
-    pub fn curve(&self) -> Curve {
-        self.0.curve
+impl<K> Zeroize for Extended<K> {
+    fn zeroize(&mut self) {
+        self.ext.zeroize()
     }
+}
 
-    pub fn extended_bytes(&self) -> &[u8; 64] {
-        self.0.extended_secret_bytes()
+impl<K> Drop for Extended<K> {
```

Comment:
Wouldn't that be done by `ZeroizeOnDrop` ?

Reply:
Good point. Let me check.